Migrate color prefs away from NSArchiver

svobs opened the pull request:
- [X] I have read [CONTRIBUTING.md](https://github.com/iina/iina/blob/develop/CONTRIBUTING.md)
- [ ] This implements/fixes issue #.

---

**Description:**

`NSArchiver` & `NSUnarchiver` are deprecated, and XCode is showing warnings. But currently IINA relies on them to serialize `NSColor` to binary format when storing the following prefs:

1. `subTextColor`
2. `subBgColor`
3. `subBorderColor`
4. `subShadowColor`

But if storing these in binary format was ever necessary, it's not now. There is already code to write the color into a string, in the format that mpv expects (search the existing code for `mpvColorString`). It wasn't too hard to change the code to store a `String` in the prefs. I did have to write a new `ValueTransformer` for it to make the XIB happy but it wasn't hard.

But changing the datatype of the 4 prefs is probably not the best idea, because doing that will cause legacy versions of IINA to show error messages when they try to read the prefs and don't get the type they were expecting.

So the code here will create 4 new pref entries with the same names but ending in `String`:
1. `subTextColorString`
2. `subBgColorString`
3. `subBorderColorString`
4. `subShadowColorString`

And all references in code are changed to use these values.

It also adds some logic at the start of IINA when the prefs are loaded:

- For the 4 prefs, it first checks to see if there is a new-style `String`-format entry.
- If `String` entry found, returns it.
- If `String` entry not found, it looks for an legacy-style `NSArchiver`-style entry.
  - If that is found, it infers that the user has just upgraded their version of IINA. So it:
    - Reads in the legacy entry,
    - Copies it into the new entry in the new `String` format
    - Returns it.
   
See `convertFromLegacyColor()` for more details.

This way, older versions of IINA will not be affected by the change to the new format, but users will be given a path away from `NSArchiver` without losing any data. Maybe after a couple versions of IINA (or when `NSArchiver` is killed by Apple), all references to the legacy entries can be dropped.  Unfortunately, until then, there will still be 1 `NSArchiver` warning (down from 4), but this seems unavoidable unless we are willing to drop the users' existing prefs for these 4 colors.

low-batt commented:
I have labeled this a high priority because this is not merely a problem with using a deprecated class. The [NSUnarchiver](https://developer.apple.com/documentation/foundation/nsunarchiver) does not provide the methods required to support [NSSecureCoding](https://developer.apple.com/documentation/foundation/nssecurecoding). That requires methods such as [unarchivedObject(ofClass:from:)](https://developer.apple.com/documentation/foundation/nskeyedunarchiver/2983380-unarchivedobject) provided by [NSKeyedUnarchiver](https://developer.apple.com/documentation/foundation/nskeyedunarchiver) where the expected class of the root object can be specified. This is about resisting object substitution attacks. Thus removing use of [NSArchiver](https://developer.apple.com/documentation/foundation/nsarchiver) is important from a security perspective. Switching to the `mpv` string based format for color (specified [here](https://mpv.io/manual/stable/#options-sub-color)) as the changes in the PR does avoids the possibility of object substitution.

I used this command to check the saved settings:
```text
defaults read com.colliderli.iina | grep 'sub.*Color.*'
```

I used these commands to force migration to repeat:
```text
defaults delete com.colliderli.iina subBgColorString
defaults delete com.colliderli.iina subBorderColorString
defaults delete com.colliderli.iina subShadowColorString
defaults delete com.colliderli.iina subTextColorString
```

low-batt commented on iina/Preference.swift:
```diff
@@ -755,16 +763,16 @@ struct Preference {
     .subTextSize: Float(55),
-    .subTextColor: NSArchiver.archivedData(withRootObject: NSColor.white),
-    .subBgColor: NSArchiver.archivedData(withRootObject: NSColor.clear),
+    .subTextColorString: convertFromLegacyColor(.subTextColorString) ?? NSColor.white.usingColorSpace(.deviceRGB)!.mpvColorString,
```
When I initially read this code it confused me. This is the `defaultPreference` property. It is supposed to contain the default value for this setting. Why would that be anything other than white in `mpv` color string form? This property is used by `AppDelegate` in a call to [register(defaults:)](https://developer.apple.com/documentation/foundation/userdefaults/1417065-register). This is establishing the _default_ values for the setting. But `convertFromLegacyColor` may return a value extracted from the original setting. Confusing. I don't think this causes a problem, but it does not match up with how [UserDefaults](https://developer.apple.com/documentation/foundation/userdefaults) should be used.

Another concern I have about this change has to do with layering and dependencies. `Preference` is one of the lowest level classes. It is even lower than `Logger` as that class must check IINA's settings to see if advanced preferences and logging are enabled. Previously there was no logging in  `Preference`. These changes add logging so now they depend upon each other. Because logging is used while forming the value for `defaultPreference` it means `Logger` is calling into `Preference` before default values have been established as can be seen in the debugger:

<details>
<summary>Stack Trace:</summary>

```text
(lldb) bt
* thread #1, queue = 'com.apple.main-thread', stop reason = breakpoint 3.1
  * frame #0: 0x000000010246edcc IINA`static Preference.bool(key=(rawValue = "enableAdvancedSettings"), self=IINA.Preference) at Preference.swift:879:12
    frame #1: 0x00000001026058f4 IINA`one-time initialization function for enabled at Logger.swift:56:35
    frame #2: 0x000000010653699c libdispatch.dylib`_dispatch_client_callout + 20
    frame #3: 0x00000001065388bc libdispatch.dylib`_dispatch_once_callout + 160
    frame #4: 0x00000001026059c8 IINA`Logger.enabled.unsafeMutableAddressor at Logger.swift:56:14
    frame #5: 0x00000001026085c4 IINA`static Logger.log(message="Converted color value from legacyKey subTextColor and stored in key subTextColorString", level=debug, subsystem=(rawValue = "iina"), appendNewlineAtTheEnd=true, self=IINA.Logger) at Logger.swift:154:11
    frame #6: 0x000000010246e08c IINA`static Preference.convertFromLegacyColor(key=(rawValue = "subTextColorString"), self=IINA.Preference) at Preference.swift:960:12
    frame #7: 0x000000010246be3c IINA`one-time initialization function for defaultPreference at Preference.swift:766:26
    frame #8: 0x000000010653699c libdispatch.dylib`_dispatch_client_callout + 20
    frame #9: 0x00000001065388bc libdispatch.dylib`_dispatch_once_callout + 160
    frame #10: 0x000000010246e334 IINA`Preference.defaultPreference.unsafeMutableAddressor at Preference.swift:679:14
    frame #11: 0x0000000102650fb8 IINA`AppDelegate.registerUserDefaultValues(self=0x00006000010f04d0) at AppDelegate.swift:755:93
    frame #12: 0x00000001026441e8 IINA`AppDelegate.applicationWillFinishLaunching(notification=Foundation.Notification @ 0x000000016d9d2220, self=0x00006000010f04d0) at AppDelegate.swift:136:5
    frame #13: 0x00000001026464e0 IINA`@objc AppDelegate.applicationWillFinishLaunching(_:) at <compiler-generated>:0
    frame #14: 0x00000001968f30f0 CoreFoundation`__CFNOTIFICATIONCENTER_IS_CALLING_OUT_TO_AN_OBSERVER__ + 148
    frame #15: 0x000000019698ee7c CoreFoundation`___CFXRegistrationPost_block_invoke + 88
    frame #16: 0x000000019698edc4 CoreFoundation`_CFXRegistrationPost + 440
    frame #17: 0x00000001968c443c CoreFoundation`_CFXNotificationPost + 704
    frame #18: 0x00000001978228a4 Foundation`-[NSNotificationCenter postNotificationName:object:userInfo:] + 88
    frame #19: 0x0000000199b171bc AppKit`-[NSApplication finishLaunching] + 248
    frame #20: 0x0000000199b16ea8 AppKit`-[NSApplication run] + 252
    frame #21: 0x0000000199aee3cc AppKit`NSApplicationMain + 880
    frame #22: 0x00000001026540a0 IINA`main at AppDelegate.swift:24:7
    frame #23: 0x00000001964c7f28 dyld`start + 2236
(lldb)
```
</details>

Another side effect of this is that these new log messages are emitted before `AppDelegate` has had a chance to emit the log file header messages:
```text
14:06:34.834 [iina][d] Converted color value from legacyKey subTextColor and stored in key subTextColorString
14:06:34.836 [iina][d] Converted color value from legacyKey subBgColor and stored in key subBgColorString
14:06:34.837 [iina][d] Converted color value from legacyKey subBorderColor and stored in key subBorderColorString
14:06:34.840 [iina][d] IINA 1.3.1 Build 133
14:06:34.840 [iina][d] Copyright © 2017-2022
14:06:34.840 [iina][d] Collider LI, et al.
14:06:34.840 [iina][d] Released under GPLv3.
```

I'm thinking the calls to `convertFromLegacyColor` should be removed from the initialization of `defaultPreference` and into a `migrateSettings` function that `AppDelegate` would then call in `applicationWillFinishLaunching` right before emitting the "App will launch" log message.

low-batt commented on iina/Preference.swift:
```diff
@@ -923,4 +927,38 @@ struct Preference {
+   Otherwise it will look for an entry for the legacy pref key. If it finds that, if will convert its value into the new format and store it under
+   the new pref key, and then return that.
+
+   This wil have the effect of automatically migrating older versions of IINA into the new format with no loss of data.
```
Typo "wil"

low-batt commented on iina/Preference.swift:
```diff
@@ -923,4 +927,38 @@ struct Preference {
+
+   This wil have the effect of automatically migrating older versions of IINA into the new format with no loss of data.
+   However, it is worth noting that this migration will only happen once, and afterwards newer versions of IINA will not look at the old pref entry.
+   And since older verions of IINA will only use the old pref entry, users who mix old and new versions of IINA may experience different values
```
Typo "verions"

low-batt commented on iina/Preference.swift:
```diff
@@ -923,4 +927,38 @@ struct Preference {
   }
 
+  /**
+   Older versions of IINA converted mpv color data into NSObject binary using the now-deprecated `NSUnarchiver` class.
```
The documentation convention is to start with a very short description of the function, followed by a blank line and then go into further discussion about the details. You can see that in Xcode if you bring up the quick help for this function. You will see that Xcode puts this entire paragraph at the top and the other paragraphs later on under a `Discussion` header.

low-batt commented on iina/Preference.swift:
```diff
@@ -923,4 +927,38 @@ struct Preference {
+   And since older verions of IINA will only use the old pref entry, users who mix old and new versions of IINA may experience different values
+   for these keys.
+   */
+  private static func convertFromLegacyColor(_ key: Key) -> String? {
```
Normally "convert" methods don't have side effects. I'm thinking "migrate" is a better term for what this function does.

low-batt commented on iina/Preference.swift:
```diff
@@ -923,4 +927,38 @@ struct Preference {
+    }
+    // Store converted value in current format for future use:
+    set(mpvColorString, for: key)
+    Logger.log("Converted color value from legacyKey \(legacyKey.rawValue) and stored in key \(key.rawValue)")
```
Same issue with the word "Converted". I think users would understand what this method is doing if the log message said "Migrated".

low-batt replied:
Might be better to say "legacy key" instead of using the variable name.

svobs commented:
Some good points. I tweaked the comments and fixed the typos, and refactored to move the code out of Preference.swift.

I don't remember the exact sequence of events which caused me to call the migration code while setting the defaults, but I agree that it's not a good design (I was probably just being hasty/sloppy). But I found I still have to call the migration code before the defaults have been set, because otherwise I don't have a reliable way to distinguish between the default values and genuine stored user data. Or at least I haven't found a way to do that. Let me know if you have any suggestions, but my workaround for dealing with any logging chicken & egg issues was to just remove the logging from the migration code.

low-batt commented:
I'm liking the new changes. Because migration of user data is a critical operation that must be solid, I am going to provide a super picky review. Please push back if you disagree with my feedback. On some of this we might want to wait and see what the other reviewers think before making changes. I will post feedback shortly.

I used this command to check the saved settings:
```text
defaults read com.colliderli.iina | grep 'sub.*Color.*\|didMigrateColorFromNSArchiver'
```

I used these commands to force migration to repeat:
```text
defaults delete com.colliderli.iina subBgColorString
defaults delete com.colliderli.iina subBorderColorString
defaults delete com.colliderli.iina subShadowColorString
defaults delete com.colliderli.iina subTextColorString
defaults delete com.colliderli.iina didMigrateColorFromNSArchiver
```

The log messages emitted at startup:
```text
18:46:01.301 [iina][d] App will launch
18:46:01.301 [iina][d] Looking for legacy color pref entries to migrate
18:46:01.302 [iina][d] Converted color value from legacy key subBgColor and stored in key subBgColorString
18:46:01.303 [iina][d] Converted color value from legacy key subBorderColor and stored in key subBorderColorString
18:46:01.304 [iina][d] Converted color value from legacy key subTextColor and stored in key subTextColorString
18:46:01.304 [iina][d] Migrated 3 of 3 legacy color pref entries
```

low-batt commented on iina/AppDelegate.swift:
```diff
@@ -198,6 +198,9 @@ class AppDelegate: NSObject, NSApplicationDelegate, SPUUpdaterDelegate {
     // register for url event
     NSAppleEventManager.shared().setEventHandler(self, andSelector: #selector(self.handleURLEvent(event:withReplyEvent:)), forEventClass: AEEventClass(kInternetEventClass), andEventID: AEEventID(kAEGetURL))
 
+    // Check for legacy pref entries and migrate them to their modern equivalents
```
Following Apple's renaming of preferences:
// Check for legacy settings and migrate them to their modern equivalents.

svobs replied:
I understand the need to use "Setting" for user-facing text, but I think it's better to continue using "Preference" internally. Mainly to keep consistency, because Preference is already being used to refer to UserDefaults entries. But I also personally like using it because "Setting" is such an overloaded term that it's not immediately clear what it's referring to without more context. A web developer could refer to cookies, local storage, or server state as "settings". And often people use the word "setting" to refer to the _UI of the setting_ instead, like with the Settings sidebar.

low-batt replied:
Sounds fine. I was rather dismayed by Apple changing this term. Imagine how much programmer time was spent on this little tweak when you add up all the time to change this in all the affected apps.

low-batt commented on iina/LegacyMigration.swift:
```diff
@@ -0,0 +1,79 @@
+
+import Foundation
+
+class LegacyMigration {
```
In Xcode this file was added under `Utils`. I'm thinking it should be under `Preference`.

svobs replied:
Good point, will change this.

low-batt commented on iina/LegacyMigration.swift:
```diff
@@ -0,0 +1,79 @@
+   for these keys.
+   */
+  static func migrateLegacyPreferences() {
+    let didMigrateColor = Preference.bool(for: .didMigrateColorFromNSArchiver)
```
When I first thought about how to do migration I thought about adding something to the settings to indicate whether settings had been migrated or not. I didn't like that because it seemed like an abuse of the intention of UserDefaults. If we are going to do this, should it be for this specific migration, or should it represent a "settings version" that we set to 2?

One change in behavior is that with the old proposed solution the migration could happen a second time if a setting was not set the first time migration occurred. If then the user ran the old version of IINA again and changed that particular color from the default, the next time the new IINA is run that setting would get migrated. This could be viewed as being better or confusing.

I think on this one we need to hear the opinion of other reviewers. At the moment I'm thinking this is fine as is.

svobs replied:
> If we are going to do this, should it be for this specific migration, or should it represent a "settings version" that we set to 2?

I too have an aversion to littering the saved state with bits like this. So much so that I didn't even consider it at first. But I don't think there's a clearly better way to record the state.

Using an incrementing version number would be a better solution if IINA was more of a closed-source / centralized project, but I think it might be limiting when considering IINA open-source nature which (at least in principle) can include multiple forks with different nonlinear offshoots which could get merged back and forth. In my mind I see parallels with how older SCM solutions like SVN used incrementing revisions, which became painful with decentralized projects. Then Git came along with its use of hashes, which turned out to be much more flexible.

There is a another possible way to go about this which I almost hesitate to mention because it would mean another round of rewrites to the PR. Instead of doing a one-time migration with a flag which gets set, we could do a two-phased migration. In the next upcoming version of IINA (and maybe some versions after, until end-of-life of `NSArchiver`) strives to keep both legacy & modern values in sync by (1) always writing to both the legacy entry and the modern entry when the user changes data, and (2) checking at startup to see if any pair of legacy-modern entries disagrees, and if so, overwrites the modern with the legacy. This would cover the situation where the user is using an older version of IINA to update their color settings, and it would always honor their most recent change. So this system of dual-entry writes would be the first phase of the migration. The second phase would occur when `NSArchiver` can no longer be used and thus the legacy entries can no longer be read, in which case that future version of IINA would read & write only to the modern entries.

But to implement that system I'd have to see how to write to both entries using Interface Builder or come up with some workaround in code. And tbh it seems like most projects don't put much effort into migrating legacy settings properly - I use Linux sometimes and I regularly see settings get reset back to defaults when upgrading, or buggy behavior that only gets solved by deleting a config file. MacOS has a poor track record with this too, and if you ever upgraded the OS and found that something stopped working properly but going into System Settings and toggling the relevant thing off and on again fixes it, you've seen a corrupted preference getting fixed. But just putting this all on the table.

low-batt replied:
I have a strong opinion on keeping the settings in sync. Normally it would be good to continue support for the old setting, but not in this case due to security. Due to the security issue with `NSArchiver` we want to remove use of the class as soon as practical. So I do not think we should add code to keep the old key in sync. We'd like to eliminate the migration code sooner than later to eliminate all use of `NSArchiver`.

Why not go back to the original scheme where migration occurs when a non-default stored setting for the old key is found and the new key does not have a persisted value? With that scheme no "migrated" flag is needed.

Might want to wait and see what the other reviewers think about this.

svobs replied:
> Why not go back to the original scheme where migration occurs when a non-default stored setting for the old key is found and the new key does not have a persisted value? With that scheme no "migrated" flag is needed.

The only problem with the previous solution is that the check needs happen before the default values are set, and thus logging is not available. I wouldn't want to wait until after the defaults are set because that would introduce a blind spot for the default value: the user may end up in a state where they want to use the default value, but the migration code will see that as a non-migration and will see their legacy value and will keep re-migrating it every time they launch.

I think the migration flag is a decent solution. The flag can double as a "receipt" for the migration, and while receipts can be annoying, they can sometimes turn out to be pretty useful.

low-batt replied:
On this:
> The only problem with the previous solution is that the check needs happen before the default values are set

That is only required when detecting if a key is persisted based on whether a value is present. Add the following to `Preference` and have `LegacyMigration` use it to check if the modern key is present in the property list.

```swift
  static func keyHasBeenPersisted(key: Key) -> Bool {
    let identifier = InfoDictionary.shared.bundleIdentifier
    guard let persisted = ud.persistentDomain(forName: identifier) else { return false }
    return persisted.keys.contains(key.rawValue)
  }
```

low-batt commented on iina/LegacyMigration.swift:
```diff
@@ -0,0 +1,79 @@
+  static func migrateLegacyPreferences() {
+    let didMigrateColor = Preference.bool(for: .didMigrateColorFromNSArchiver)
+    guard !didMigrateColor else {
+      Logger.log("Already migrated color prefs away from NSArchiver; no migration needed", level: .verbose)
```
```suggestion
      Logger.log("Already migrated color settings away from NSArchiver; no migration needed", level: .verbose)
```

low-batt commented on iina/LegacyMigration.swift:
```diff
@@ -0,0 +1,79 @@
+
+class LegacyMigration {
+
+  // Map of [modern pref key → legacy key]. Do not reference legacy keys outside this file.
```
```suggestion
  /// Map of [modern settings key → legacy key].
  ///
  /// - Important: Do not reference legacy keys outside this file.
```

low-batt commented on iina/LegacyMigration.swift:
```diff
@@ -0,0 +1,79 @@
+   To do this in a way which does not corrupt the values for older versions of IINA, we'll store the new format under a new `Preference.Key`,
+   and leave the legacy pref entry as-is.
+
+   This method will be executed on each of the affected prefs when IINA starts up. It will first check if there is already an entry for the new
```
This reflects the original PR code and does not match what the code is doing now.

low-batt replied:
This was not updated and still reflects the original code which did not use `didMigrateColorFromNSArchiver`. Of course I'm thinking maybe we should go back to that method. If we aren't doing that and sticking with keying off `didMigrateColorFromNSArchiver` then this description of what the method does needs to be updated:
```text
   This method will be executed on each of the affected prefs when IINA starts up. It will first check if there is already an entry for the new
   pref key. If it finds one, then it will assume that the migration has already occurred, and will just return that.
   Otherwise it will look for an entry for the legacy pref key. If it finds that, if will convert its value into the new format and store it under
   the new pref key, and then return that.

low-batt commented on iina/LegacyMigration.swift:
```diff
@@ -0,0 +1,79 @@
+   And since older versions of IINA will only use the old pref entry, users who mix old and new versions of IINA may experience different values
+   for these keys.
+   */
+  static func migrateLegacyPreferences() {
```
migrateLegacySettings following Apple's new naming scheme.

low-batt commented on iina/LegacyMigration.swift:
```diff
@@ -0,0 +1,79 @@
+      return
+    }
+
+    Logger.log("Looking for legacy color pref entries to migrate")
```
```suggestion
    Logger.log("Looking for legacy color settings to migrate")
```

low-batt commented on iina/LegacyMigration.swift:
```diff
@@ -0,0 +1,79 @@
+    var entriesMigratedCount: Int = 0
+    for (modernKey, legacyKey) in legacyColorPrefKeyMap {
+      // Look for legacy pref:
+      guard let data = UserDefaults.standard.data(forKey: legacyKey.rawValue) else { continue }
```
```suggestion
      guard let data = Preference.data(for: legacyKey) else { continue }
```

low-batt commented on iina/LegacyMigration.swift:
```diff
@@ -0,0 +1,79 @@
+        return  // Data error == fail migration
+      }
+      // Store migrated value in modern string format under modern pref key:
+      UserDefaults.standard.set(mpvColorString, forKey: modernKey.rawValue)
```
```suggestion
      Preference.set(mpvColorString, for: modernKey)
```

low-batt commented on iina/LegacyMigration.swift:
```diff
@@ -0,0 +1,79 @@
+      Logger.log("Converted color value from legacy key \(legacyKey.rawValue) and stored in key \(modernKey.rawValue)")
+      entriesMigratedCount += 1
+    }
+    Logger.log("Migrated \(entriesMigratedCount) of \(entriesFoundCount) legacy color pref entries")
```
```suggestion
    Logger.log("Migrated \(entriesMigratedCount) of \(entriesFoundCount) legacy color settings")
```

low-batt commented on iina/LegacyMigration.swift:
```diff
@@ -0,0 +1,79 @@
+      // A pref which has not been set looks identical to the default value.
+      let defaultValue = Preference.defaultPreference[modernKey] as? String
+      let currentValue = Preference.value(for: modernKey) as? String
+      guard currentValue == nil || currentValue == defaultValue else {
```
```suggestion
      let currentValue = Preference.string(for: modernKey)!
      guard currentValue == defaultValue else {
```

low-batt reviewed (commented):
Looking good. Some minor tweaks.

svobs commented:
> That is only required when detecting if a key is persisted based on whether a value is present. Add the following to Preference and have LegacyMigration use it to check if the modern key is present in the property list.

Great tip - thanks. I went ahead and implemented this and removed the `didMigrateColorFromNSArchiver` flag, and it's much cleaner.

uiryuu commented on iina/LegacyMigration.swift:
```diff
@@ -0,0 +1,72 @@
+   for these keys.
+   */
+  static func migrateLegacyPreferences() {
+    Logger.log("Looking for legacy color prefs to migrate", level: .verbose)
```
I think we should make this the same log level as other logs, there's no reason for only setting this verbose.

uiryuu commented on iina/LegacyMigration.swift:
```diff
@@ -0,0 +1,72 @@
+      }
+      // Store string under modern pref key:
+      Preference.set(mpvColorString, for: modernKey)
+      Logger.log("Converted color value from legacy pref \(legacyKey.rawValue) and stored in pref \(modernKey.rawValue)")
```
Should we remove the legacy entry after migration?

low-batt replied:
Not now. Especially not in a beta. With a beta release it is much more likely users will go back and forth between releases.

uiryuu commented on iina/Preference.swift:
```diff
@@ -959,4 +955,9 @@ struct Preference {
 
+  static func keyHasBeenPersisted(_ key: Key) -> Bool {
+    let identifier = InfoDictionary.shared.bundleIdentifier
+    guard let persisted = ud.persistentDomain(forName: identifier) else { return false }
```
I suggest make it to be
``` swift
static let persisted = ud.persistentDomain(forName: identifier)
guard let _ = persisted else { return false }
```
to prevent retrieving the whole UserDefault multiple times.

lhc70000 replied:
If so, `persisted` will only be updated when IINA starts. If someone find this function in the future and try to use it, they may encounter problems if they updated the preference when IINA is running. I actually suggest moving this part to `LegacyMigration` where it's being used, as an inner function or block.

uiryuu reviewed (changes requested):
Generally looks good, here're some suggestions

svobs committed:
Add auto migration of stored color prefs from NSArchiver data to strings

svobs commented:
Rebased and refactored based on above feedback.

uiryuu reviewed (approved):
Changes look good to me
